### DISCUSSION TITLE
Dependency injection on Generated StateNotifierProvider

### DISCUSSION THREAD

**Discussion by AslamThachapalli** [rrousselGit/riverpod#3280] [Q&A]
- What will be the ideal approach if we want to do dependency injection for the generated (Riverpod 2.0) StateNotifierProvider?
- The way I am doing it is:

```dart

// The StateNotifier.
@riverpod
class LoginStateNotifier extends _$LoginStateNotifier {
  @override
  LoginState build(LoginRepository repository) {
    return const LoginState.initial();
  }
}

// Provider registration. I suppose mingling generated and ungenerated providers is not a good practice.
// Lets assume this is a generated provider for now. 
final loginNotifierProvider = Provider((ref) {
  final repository = ref.watch(loginRepoProvider);

  return loginStateNotifierProvider(repository);
});

// How I call it:
ref
    .read(ref.watch(loginNotifierProvider).notifier)
    .method();
```

*Also I am facing difficulty in trying to mock this as well.* Is there any good resource to understand unit testing with riverpod.

Is this a recommended way? I feel it a bit overdo as I liked the ungenerated way of approach for the dependency injection.

- Or should I continue with the ungenerated way?  Will it affect the project in a long run as riverpod team is favoring code generation approach more?

**Comment by rrousselGit** [comment]
For testing practices, in particular mocking Notifiers, see https://riverpod.dev/docs/essentials/testing#mocking-notifiers

**Reply by AslamThachapalli** [reply]
Thanks
